src/eventer/eventer_jobq.c

```diff
@@ -327,19 +418,46 @@ eventer_jobq_enqueue(eventer_jobq_t *jobq, eventer_job_t *job, eventer_job_t *pa
 static eventer_job_t *
 __eventer_jobq_dequeue(eventer_jobq_t *jobq, int should_wait) {
   eventer_job_t *job = NULL;
+  int cycles = 0;
 
   /* Wait for a job */
   if(should_wait) while(sem_wait(&jobq->semaphore) && errno == EINTR);
   /* Or Try-wait for a job */
   else if(sem_trywait(&jobq->semaphore)) return NULL;
 
   pthread_mutex_lock(&jobq->lock);
-  if(jobq->headq) {
-    /* If there are items, pop and advance the header pointer */
-    job = jobq->headq;
-    jobq->headq = jobq->headq->next;
-    if(!jobq->headq) jobq->tailq = NULL;
+  eventer_jobsq_t *starting_point = jobq->current_squeue;
+  uint32_t tgt_inflight = 0;
+  /* We're going to spin around our work queues aiming for a balance
+   * of inflight jobs per queue.
+   * First choose the next queue with <= (concurrent/queues) inflight jobs.
+   * There are possible rounding errors, so next bump by one and repeat.
+   * If we have no jobs it means some of the queues have no more work,
+   * so we run one last time with no inflight limit.
+   */
+  while(job == NULL) {
+    eventer_jobsq_t *squeue = jobq->current_squeue;
+    if(squeue == starting_point) {
+      cycles++;
+      tgt_inflight = (jobq->concurrency / jobq->subqueue_count);
+      if(cycles == 1 && tgt_inflight == 0) {
+        cycles = 2; tgt_inflight = 1;
+      }
+
+      if(cycles == 1) tgt_inflight--;
+      else if(cycles == 2) (void)tgt_inflight; /* no op */
+      else if(cycles == 3) tgt_inflight = UINT32_MAX;
+      else break;
+    }
+    if(squeue->headq && squeue->inflight <= tgt_inflight) {
+      /* If there are items, pop and advance the header pointer */
+      job = squeue->headq;
+      squeue->headq = squeue->headq->next;
+      if(!squeue->headq) squeue->tailq = NULL;
+    }
+    jobq->current_squeue = jobq->current_squeue->next;
```

Comment:
We advance the current_squeue here but never deal with the end of the list (NULL)?

Reply:
It's a circulary queue... jobq->queue.next = prev = self in initialization.